### ISSUE TITLE
Parse failure for 'if' with flags inside test-suite

### ISSUE THREAD

**Issue by ChShersh** [haskell/cabal#5055]
I'm following these suggestions from official documentation to use different `main` files for different flags:

* https://www.haskell.org/cabal/users-guide/developing-packages.html#meaning-of-field-values-when-using-conditionals

> **Real problem:** I need to disable `doctest` tests for Windows because AppVeyor CI fails

But when I'm trying to use this approach for my test-suite I see parsing error when using `cabal new-build` command.

#### Expected behavior

Build without problems

#### Observed behavior

```
$ cabal new-build
Warning:
.../cabal-flag-example/cabal-flag-example.cabal:11:1:
The 'main-is' field is required for the exitcode-stdio-1.0 test suite type.
cabal: Failed parsing
".../cabal-flag-example/./cabal-flag-example.cabal".
```

I've tried to create minimal reproducible example. You can see it here:

* https://github.com/ChShersh/cabal-flag-example

```shell
$ cabal --version
cabal-install version 2.1.0.0
compiled using version 2.1.0.0 of the Cabal library 
```

**Comment by phadej** [comment]
So far I think this is a documentation bug

Trying with `cabal-install-2.0` gives:

```
% /opt/cabal/2.0/bin/cabal check
cabal: ./cabal-flag-example.cabal:11: The 'main-is' field is required for the
exitcode-stdio-1.0 test suite type.
```

For now you can workaround this by writing (conditional statements are interpreted after "main" body):

```
test-suite flag-cabal-test
  type:                exitcode-stdio-1.0
  build-depends:       base >= 4.8 && < 5
  default-language:    Haskell2010

  main-is:             SecondMain.hs

  if os(windows)
    main-is:             FirstMain.hs
```

or having dummy `Main.hs` which includes module based on `hs-source-dirs`:

```haskell
module Main (main) where
import Tests (main)
```

and

```
test-suite flag-cabal-test
  type:                exitcode-stdio-1.0
  build-depends:       base >= 4.8 && < 5
  default-language:    Haskell2010

  main-is: Main.hs

  hs-source-dirs: common
  other-modules: Tests

  if os(windows)
    hs-source-dirs: windows
  else 
    hs-source-dirs: unix
```

---

Currently we are conservative when checking that test-suite always have `type`, and if it's `exitcode-stdio-1.0` then also `main-is` (for `detailed-0.9` we required `test-module`). To fix this we need to verify that all possible "flattenings" have `type`, and then either `main-is` or `test-module` depending on the `type`. More concretely we need to generalise `validateTestSuite` check into `validateCondTreeTestSuite`. 

Now as I look into the code, that check looks fishy anyway as

```
test-suite flag-cabal-test
  build-depends:       base >= 4.8 && < 5
  default-language:    Haskell2010

  if os(windows)
    type:                exitcode-stdio-1.0
    main-is:             FirstMain.hs
  else
    type:                exitcode-stdio-1.0
    main-is:             SecondMain.hs
```

works

**AND** with current-ish `master` also

```
test-suite flag-cabal-test
  build-depends:       base >= 4.8 && < 5
  default-language:    Haskell2010

  if os(windows)
    type:                exitcode-stdio-1.0
    main-is:             FirstMain.hs
```

which should fail (and fails with `cabal-install-2.0`).

**Comment by ChShersh** [comment]
@phadej Thanks a lot for your response! Your workarounds really help me :+1:

**Comment by 23Skidoo** [comment]
Fixed by #5076.
